Userland/Applications/Escalator/EscalatorWindow.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2022, Ashley N. <dev-serenity@ne0ndrag0n.com>
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "EscalatorWindow.h"
+#include <AK/Assertions.h>
+#include <Applications/Escalator/EscalatorGML.h>
+#include <LibCore/File.h>
+#include <LibCore/SecretString.h>
+#include <LibCore/System.h>
+#include <LibGUI/FileIconProvider.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/Widget.h>
+#include <unistd.h>
+
+EscalatorWindow::EscalatorWindow(StringView description, Vector<StringView> command, Core::Account current_user)
+    : m_command(command)
+    , m_current_user(current_user)
+{
+    auto app_icon = find_icon();
+
+    set_title("Run as Root");
+    set_icon(app_icon.bitmap_for_size(16));
+    resize(345, 100);
+    set_resizable(false);
+    set_minimizable(false);
+
+    auto& main_widget = set_main_widget<GUI::Widget>();
+    main_widget.load_from_gml(escalator_gml);
+
+    RefPtr<GUI::Label> app_label = *main_widget.find_descendant_of_type_named<GUI::Label>("description");
+
+    String prompt;
+    if (description.is_empty())
+        prompt = String::formatted("{} requires root access. Please enter password for user \"{}\".", command[0], m_current_user.username());
+    else
+        prompt = description;
+
+    app_label->set_text(prompt);
+
+    m_icon_image_widget = *main_widget.find_descendant_of_type_named<GUI::ImageWidget>("icon");
+    m_icon_image_widget->set_bitmap(app_icon.bitmap_for_size(32));
+
+    m_ok_button = *main_widget.find_descendant_of_type_named<GUI::DialogButton>("ok_button");
+    m_ok_button->on_click = [this](auto) {
+        auto result = check_password();
+        if (result.is_error()) {
+            GUI::MessageBox::show_error(this, String::formatted("Failed to execute command: {}", result.error()));
+            close();
+        }
+    };
+    m_ok_button->set_default(true);
+
+    m_cancel_button = *main_widget.find_descendant_of_type_named<GUI::DialogButton>("cancel_button");
+    m_cancel_button->on_click = [this](auto) {
+        close();
+    };
+
+    m_password_input = *main_widget.find_descendant_of_type_named<GUI::PasswordBox>("password");
+}
+
+GUI::Icon EscalatorWindow::find_icon()
+{
+    auto executable_path = Core::File::resolve_executable_from_environment(m_command[0]);
+
+    if (executable_path.has_value())
+        return GUI::FileIconProvider::icon_for_executable(executable_path.value());
+
+    return GUI::Icon::default_icon("app-escalator"sv);
```

Comment:
This logic looks a bit weird. In a case where we aren't able to resolve an executable path, the `exec` will presumably fail to run anyways. The more interesting case (for the purpose of returning an icon) would likely be if there is no special icon for that executable (i.e. we'd have to specify that as a fallback for `FileIconProvider`).

Reply:
Updated to check once and pass to the window if valid. Error will now be thrown if the command fails to resolve and returns 127 instead of showing the window.